runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,482 @@
+end
+
+--- Computes the 0-based position of a tabstop located at the end of `snippet` and spanning
+--- `placeholder` (if given).
+---
+--- @param snippet string[]
+--- @param placeholder string?
+--- @return Range4
+local function compute_tabstop_range(snippet, placeholder)
+  local cursor_row, cursor_col = cursor_pos()
+  local snippet_text = remove_newlines(snippet)
+  local placeholder_text = remove_newlines(placeholder or '')
+  local start_row = cursor_row + #snippet_text - 1
+  local start_col = #(snippet_text[#snippet_text] or '')
+
+  -- Add the cursor's column offset to the first line.
+  if start_row == cursor_row then
+    start_col = start_col + cursor_col
+  end
+
+  return {
+    start_row,
+    start_col,
+    start_row + #placeholder_text - 1,
+    start_col + #(placeholder_text[#placeholder_text] or ''),
+  }
+end
+
+--- @class vim.snippet.Tabstop
+--- @field extmark_id integer
+--- @field index integer
+--- @field bufnr integer
+local Tabstop = {}
+Tabstop.__index = Tabstop
+
+--- Creates a new tabstop.
+---
+--- @param index integer
+--- @param bufnr integer
+--- @param range Range4
+--- @return vim.snippet.Tabstop
+function Tabstop.new(index, bufnr, range)
+  local extmark_id = vim.api.nvim_buf_set_extmark(bufnr, snippet_ns, range[1], range[2], {
+    right_gravity = false,
+    end_right_gravity = true,
+    end_line = range[3],
+    end_col = range[4],
+  })
+
+  local self = setmetatable(
+    { index = index, bufnr = bufnr, extmark_id = extmark_id },
+    { __index = Tabstop }
+  )
+
+  return self
+end
+
+--- Returns the tabstop's range.
+---
+--- @return Range4
+function Tabstop:get_range()
+  local mark =
+    vim.api.nvim_buf_get_extmark_by_id(self.bufnr, snippet_ns, self.extmark_id, { details = true })
+
+  --- @diagnostic disable-next-line: undefined-field
+  return { mark[1], mark[2], mark[3].end_row, mark[3].end_col }
+end
+
+--- Returns the text spanned by the tabstop.
+---
+--- @return string
+function Tabstop:get_text()
+  local range = self:get_range()
+  local lines = vim.api.nvim_buf_get_lines(self.bufnr, range[1], range[3] + 1, false)
+  lines[#lines] = lines[#lines]:sub(1, range[4])
+  lines[1] = lines[1]:sub(range[2] + 1)
+
+  return table.concat(lines, '\n')
+end
+
+--- Sets the tabstop's text.
+---
+--- @param text string
+function Tabstop:set_text(text)
+  local range = self:get_range()
+  vim.api.nvim_buf_set_text(
+    self.bufnr,
+    range[1],
+    range[2],
+    range[3],
+    range[4],
+    remove_newlines(text)
+  )
+end
+
+--- @class vim.snippet.Session
+--- @field bufnr integer
+--- @field tabstops table<integer, vim.snippet.Tabstop[]>
+--- @field current_tabstop vim.snippet.Tabstop
+local Session = {}
+
+--- Initializes a new snippet session in the current buffer.
+---
+--- @return vim.snippet.Session
+function Session:init()
+  self.bufnr = vim.api.nvim_get_current_buf()
+  self.tabstops = {}
+  self.current_tabstop = Tabstop.new(0, self.bufnr, { 0, 0, 0, 0 })
+
+  return self
+end
```

Comment:
Is this intentionally mutating the `Session` object?
If not:


```suggestion
function Session.new()
  local state = {
    bufnr = vim.api.nvim_get_current_buf(),
    tabstops = {},
    current_tabstop = Tabstop.new(0, self.bufnr, { 0, 0, 0, 0 })
  }
  return setmetatable(state, __index = Session)
end
```

If it is intentional, I'd tend towards using a regular global table, without class style because class style usually involves different instances with encapsulated state.

Reply:
Yeah I think I'll just get rid of the `Session` class and move it's functionality to `M` as private bits.

Reply:
One thing to keep in mind is that we may want to allow plugins to hook into the "expand and jump" functionality -- maybe creating a Session would be one way?

Reply:
That's a good point, it does allow for better extensibility in the future. I'll create session instances instead then.